### PR TITLE
Examples lottery esdt

### PR DESCRIPTION
**Pull request by dorin-iancu**

### CODE REVIEW

**Review by andrei-marinica — APPROVED** [review]

File: examples/lottery-esdt/src/random.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+		if self.current_index == SEED_SIZE {
+			self.shuffle();
+			self.current_index = 0;
+		}
+
+		return first_byte | second_byte | third_byte | fourth_byte;
+	}
+
+	// Fake shuffle. Just add numbers to one another, accounting for overflow overflow.
```
Comment (sasurobert) [inline examples/lottery-esdt/src/random.rs:32]: overflow overflow ?

File: examples/lottery-esdt/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+		lottery_name: Vec<u8>,
+		esdt_token_name: Vec<u8>,
+		ticket_price: BigUint,
+		opt_total_tickets: Option<u32>,
+		opt_deadline: Option<u64>,
+		opt_max_entries_per_user: Option<u32>,
+		opt_prize_distribution: Option<Vec<u8>>,
+		opt_whitelist: Option<Vec<Address>>,
+	) -> SCResult<()> {
+		self.start_lottery(
```
Comment (sasurobert) [inline examples/lottery-esdt/src/lib.rs:62]: this endpoint does exactly the same call as the start. do we need both ?

File: examples/lottery-esdt/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+				self.clear_storage(&lottery_name);
+				Ok(())
+			},
+		}
+	}
+
+	#[view]
+	fn status(&self, lottery_name: Vec<u8>) -> Status {
+		let exists = self.get_lottery_exists(&lottery_name);
+
```
Comment (sasurobert) [inline examples/lottery-esdt/src/lib.rs:181]: delete empty line

File: examples/lottery-esdt/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+		if !exists {
+			return Status::Inactive;
+		}
+
+		let info = self.get_mut_lottery_info(&lottery_name);
+
```
Comment (sasurobert) [inline examples/lottery-esdt/src/lib.rs:187]: delete empty line

File: examples/lottery-esdt/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+		let mut info = self.get_mut_lottery_info(&lottery_name);
+		let caller = self.get_caller();
+		let call_token_name = self.get_esdt_token_name();
+		let payment = self.get_esdt_value_big_uint();
+
+		require!(
+			info.whitelist.is_empty() || info.whitelist.contains(&caller),
+			"You are not allowed to participate in this lottery!"
+		);
+
```
Comment (sasurobert) [inline examples/lottery-esdt/src/lib.rs:205]: delete empty line

File: examples/lottery-esdt/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+		require!(call_token_name == info.esdt_token_name, "Wrong esdt token!");
+
```
Comment (sasurobert) [inline examples/lottery-esdt/src/lib.rs:207]: delete empty line

File: examples/lottery-esdt/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+		require!(payment == info.ticket_price, "Wrong ticket fee!");
+
+		let mut entries = self.get_mut_number_of_entries_for_user(&lottery_name, &caller);
+
```
Comment (sasurobert) [inline examples/lottery-esdt/src/lib.rs:211]: delete empty line

File: examples/lottery-esdt/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+				for_loop_end = info.prize_distribution.len();
+			}
+
+			// distribute to the first place last. Laws of probability say that order doesn't matter.
+			// this is done to mitigate the effects of BigUint division leading to "spare" prize money being left out at times
+			// 1st place will get the spare money instead.
+			for i in (0..for_loop_end).rev() {
+				let mut winning_ticket_id: u32;
+
+				loop {
```
Comment (sasurobert) [inline examples/lottery-esdt/src/lib.rs:253]: can you take this out in a separate function ?

File: examples/lottery-esdt/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+		let mut serializer = CallDataSerializer::new(ESDT_TRANSFER_STRING);
+		serializer.push_argument_bytes(esdt_token_name);
+		serializer.push_argument_bytes(amount.to_bytes_be().as_slice());
+
+		self.async_call(&to, &BigUint::zero(), serializer.as_slice());
+	}
+
+	fn sum_array(&self, array: &[u8]) -> u16 {
+		let mut sum = 0u16; // u16 to protect against overflow
+
```
Comment (sasurobert) [inline examples/lottery-esdt/src/lib.rs:327]: delete empty line

**Pull request merged**